src/xen/xennet.c

```diff
@@ -508,56 +564,84 @@ static void xennet_populate_rx_ring(xennet_dev xd)
         xen_notify_evtchn(xd->evtchn);
 }
 
-closure_function(2, 0, void, xennet_rx_buf_finish,
-                 xennet_dev, xd, struct pbuf *, p)
-{
-    xennet_dev xd = bound(xd);
-    struct pbuf *p = bound(p);
-    if (xd->netif->input(p, xd->netif) != ERR_OK) {
-        msg_err("rx drop by stack\n");
-        xennet_return_rxbuf(p);
-    }
-    closure_finish();
-}
-
 static void xennet_service_rx_ring(xennet_dev xd)
 {
     int more;
 
     do {
         RING_IDX cons = xd->rx_ring.rsp_cons;
         RING_IDX prod = xd->rx_ring.sring->rsp_prod;
-        assert(prod - cons < XENNET_RX_RING_SIZE);
+        assert(prod - cons <= XENNET_RX_RING_SIZE);
         read_barrier();
         xennet_debug("%s: cons %d, prod %d", __func__, cons, prod);
 
+        struct list q;
+        list_init(&q);
+
+        /* again, lock unfortunately needed for rxbufs - wouldn't be
+           an issue if we could size the vector only once on init... */
+        u64 flags = spin_lock_irq(&xd->rx_fill_lock);
```

Comment:
In that particular case it's because the rxbufs vector could be resized on rx buffer alloc... but in general, yes, it's for exclusion between a ring fill happening from the kernel context and a ring fill at the end of interrupt service.